lib/Reader/readerir.cpp

```diff
@@ -4739,12 +4774,57 @@ bool GenIR::canMakeDirectCall(ReaderCallTargetData *CallTargetData) {
   return !CallTargetData->isJmp();
 }
 
-IRNode *GenIR::makeDirectCallTargetNode(void *CodeAddr) {
-  uint32_t NumBits = TargetPointerSizeInBits;
-  bool IsSigned = false;
+GlobalVariable *GenIR::GetGlobalVariable(uint64_t LookupHandle,
+                                         uint64_t ValueHandle, Type *Ty,
+                                         StringRef Name, bool IsConstant) {
+  GlobalVariable *GlobalVar = nullptr;
+  auto MapElement = HandleToGlobalVariableMap.find(LookupHandle);
+  if (MapElement != HandleToGlobalVariableMap.end()) {
+    GlobalVar = MapElement->second;
+  }
+
+  if (GlobalVar == nullptr) {
+    Constant *Initializer = nullptr;
+    const bool IsExternallyInitialized = false;
+    const GlobalValue::LinkageTypes LinkageType =
+        GlobalValue::LinkageTypes::ExternalLinkage;
+    GlobalVariable *const InsertBefore = nullptr;
+    unsigned int AddressSpace = 0;
+
+    GlobalVar = new GlobalVariable(*JitContext->CurrentModule, Ty, IsConstant,
+                                   LinkageType, Initializer, Name, InsertBefore,
+                                   GlobalValue::NotThreadLocal, AddressSpace,
+                                   IsExternallyInitialized);
+    StringRef ResultName = GlobalVar->getName();
+    if (NameToHandleMap->count(ResultName) == 1) {
+      assert((*NameToHandleMap)[ResultName] == ValueHandle);
+    } else {
+      (*NameToHandleMap)[ResultName] = ValueHandle;
+    }
+    HandleToGlobalVariableMap[LookupHandle] = GlobalVar;
+  }
+
+  return GlobalVar;
+}
 
-  ConstantInt *CodeAddrValue = ConstantInt::get(
-      *JitContext->LLVMContext, APInt(NumBits, (uint64_t)CodeAddr, IsSigned));
+IRNode *GenIR::makeDirectCallTargetNode(CORINFO_METHOD_HANDLE MethodHandle,
+                                        mdToken MethodToken, void *CodeAddr) {
+  uint64_t Handle = (uint64_t)CodeAddr;
+  Type *CodeAddrTy =
+      Type::getIntNTy(*JitContext->LLVMContext, TargetPointerSizeInBits);
+  const bool IsConstant = true;
+  const char *ModuleName = nullptr;
+  const char *MethodName =
+      JitContext->JitInfo->getMethodName(MethodHandle, &ModuleName);
+
+  std::string FullName;
+  raw_string_ostream OS(FullName);
+  OS << format("%s.%s(TK_%x)", ModuleName, MethodName, MethodToken);
+  OS.flush();
+  GlobalVariable *GlobalVar =
+      GetGlobalVariable(Handle, Handle, CodeAddrTy, FullName, IsConstant);
```

Comment:
Should we be using `Function`s instead of `GlobalVariable`s for these?  (I'm not trying to say we should, just wondering what your thoughts are.  On the one hand, with Functions we could apply function-level attributes that analyses might pick up on; on the other hand, these really are pretty opaque to us so maybe it's safer to model them as indirect calls via function pointer global variables.)